worker/peergrouper/desired.go

```diff
@@ -192,6 +201,57 @@ func desiredPeerGroup(info *peerGroupInfo) (map[string]*replicaset.Member, map[s
 	}
 	return newMembers, machineVoting, nil
 }
+// reviewPeerGroupChanges adds some extra logic after creating possiblePeerGroupChanges to safely add
+// or remove machines keeping the correct odd number of voters peer structure, and preventing the
+// primary peer demoting.
+func reviewPeerGroupChanges(newMembers map[string]*replicaset.Member,
+							info *peerGroupInfo,
+							peerChanges *peerGroupChanges){
+
+	currVoters := 0
+	for _, m := range newMembers {
+		if isVotingMember(m) {
+			currVoters += 1
+		}
+	}
+	keptVoters := currVoters - len(peerChanges.toRemoveVote)
+	if (keptVoters+len(peerChanges.toAddVote))%2 != 1 {
+		// if this is true we will create an even number of voters
+
+		if len(peerChanges.toAddVote) > 0 {
+			peerChanges.toAddVote = peerChanges.toAddVote[:len(peerChanges.toAddVote)-1]
```

Comment:
we can similarly return after this, rather than nesting the else, because we know we now have a safe number of peers.

I think we need to mutate "toKeep" again.